File: tools/math_dsl/state_machine.py
"""W244 wave 17 — closed-form analytical model for `state_machine` (supermeter).

Industry pattern (Stakelogic Supermeter, Pragmatic Power of Thor mode switch,
Big Bass Splash multi-mode FS, Aristocrat Buffalo Stampede tier escalation):

  Slot game has a persistent state (e.g. `base | super | mega`) that
  transitions according to a Markov chain. Each state has:

    * Its own per-spin RTP component (different paytable / wild weights)
    * Outgoing transitions to other states with probability `p_transition`

  Examples:
    * Supermeter: base state with rare upgrade to super mode (3× RTP),
      eventually drops back to base after configured spins.
    * Mode switch: base / fury mode toggle on landed scatter; fury has
      higher hit_freq but lower max win.

  Closed-form RTP contribution
  ----------------------------
    Stacionary distribution of the Markov chain × per-state RTP, by
    ergodic theorem:

        RTP_total = sum_states(π_s × rtp_s)

    where π is the stationary distribution of the transition matrix P:

        π = π × P   (left-eigenvector with eigenvalue 1)
        sum(π) = 1

  Computation
  -----------
    For small state spaces (3-5 states typical), we solve the linear
    system directly: (P^T - I) π = 0  with normalization sum(π) = 1.

    Pure-stdlib Gaussian elimination (no numpy dependency).

Pure-stdlib. Used by:
  * `tools.math_dsl.compile` for IR emission
  * `tools/build_state_machine_kernel.py` for acceptance artefact
  * `tools/tests/test_w244_state_machine_kernel.py` for closed-form pin
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class GameState:
    """One state of the slot's state machine."""
    name: str                        # "base" | "super" | "mega" | "fury"
    rtp_component: float             # per-spin RTP when in this state

    def __post_init__(self):
        if not self.name:
            raise ValueError("state name must be non-empty")
        if self.rtp_component < 0:
            raise ValueError("rtp_component must be ≥ 0")


@dataclass(frozen=True)
class StateMachineParams:
    """Closed-form model inputs."""
    states: tuple[GameState, ...]
    # Transition matrix: transitions[from_idx][to_idx] = probability.
    # Rows must sum to 1.0 (stochastic).
    transitions: tuple[tuple[float, ...], ...]

    def __post_init__(self):
        n = len(self.states)
        if n == 0:
            raise ValueError("states must be non-empty")
        if len(self.transitions) != n:
            raise ValueError(
                f"transitions has {len(self.transitions)} rows, expected {n}"
            )
        for i, row in enumerate(self.transitions):
            if len(row) != n:
                raise ValueError(
                    f"transitions row {i} has {len(row)} entries, expected {n}"
                )
            row_sum = sum(row)
            if abs(row_sum - 1.0) > 1e-9:
                raise ValueError(
                    f"transitions row {i} sums to {row_sum}, expected 1.0"
                )
            if any(p < 0 for p in row):
                raise ValueError(f"transitions row {i} has negative probability")


def stationary_distribution(params: StateMachineParams) -> tuple[float, ...]:
    """Solve π × P = π with sum(π) = 1 via Gaussian elimination on (P^T - I).

    Replaces last row with sum-to-1 constraint and RHS = [0, ..., 0, 1].
    """
    n = len(params.states)
    # Build augmented matrix A | b where A is (P^T - I) with last row replaced.
    A = [[0.0] * n for _ in range(n)]
    b = [0.0] * n
    for i in range(n - 1):
        for j in range(n):
            A[i][j] = params.transitions[j][i]  # P^T element
        A[i][i] -= 1.0
        # b[i] = 0 already
    # Last row: sum(π) = 1
    for j in range(n):
        A[n - 1][j] = 1.0
    b[n - 1] = 1.0

    # Gaussian elimination with partial pivoting
    for i in range(n):
        # Find pivot
        pivot_row = i
        max_val = abs(A[i][i])
        for k in range(i + 1, n):
            if abs(A[k][i]) > max_val:
                max_val = abs(A[k][i])
                pivot_row = k
        if max_val < 1e-15:
            raise ValueError(
                "Transition matrix is singular (no unique stationary distribution)"
            )
        # Swap rows
        A[i], A[pivot_row] = A[pivot_row], A[i]
        b[i], b[pivot_row] = b[pivot_row], b[i]
        # Eliminate below
        for k in range(i + 1, n):
            factor = A[k][i] / A[i][i]
            for j in range(i, n):
                A[k][j] -= factor * A[i][j]
            b[k] -= factor * b[i]

    # Back-substitution
    pi = [0.0] * n
    for i in range(n - 1, -1, -1):
        s = b[i]
        for j in range(i + 1, n):
            s -= A[i][j] * pi[j]
        pi[i] = s / A[i][i]

    # Numerical cleanup: clamp tiny negatives to 0 + renormalize
    pi = [max(0.0, x) for x in pi]
    total = sum(pi)
    if total <= 0:
        raise ValueError("computed stationary distribution sums to ≤ 0")
    pi = [x / total for x in pi]
    return tuple(pi)


def state_machine_rtp(params: StateMachineParams) -> dict:
    """Per-spin RTP contribution + per-state breakdown."""
    pi = stationary_distribution(params)
    per_state = []
    total_rtp = 0.0
    for s, prob in zip(params.states, pi):
        contrib = prob * s.rtp_component
        per_state.append({
            "name": s.name,
            "stationary_probability": prob,
            "rtp_component_in_state": s.rtp_component,
            "weighted_rtp_contribution": contrib,
        })
        total_rtp += contrib
    return {
        "rtp_contribution": total_rtp,
        "states_count": len(params.states),
        "stationary_distribution": list(pi),
        "states": per_state,
    }

File: rust-sim/src/kernels/state_machine.rs
//! W244.17 — state_machine Rust port.
//!
//! Markov stationary distribution via Gaussian elimination on (P^T - I)
//! sa sum=1 last-row constraint. Pure Rust, no numpy/scipy.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameState {
    pub name: String,
    pub rtp_component: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StateMachineParams {
    pub states: Vec<GameState>,
    /// Transition matrix: transitions[from][to] = probability.
    /// Each row must sum to 1.0 (stochastic).
    pub transitions: Vec<Vec<f64>>,
}

impl StateMachineParams {
    pub fn validate(&self) -> Result<(), String> {
        let n = self.states.len();
        if n == 0 {
            return Err("states must be non-empty".into());
        }
        if self.transitions.len() != n {
            return Err(format!(
                "transitions has {} rows, expected {}",
                self.transitions.len(), n
            ));
        }
        for (i, row) in self.transitions.iter().enumerate() {
            if row.len() != n {
                return Err(format!(
                    "transitions row {} has {} entries, expected {}",
                    i, row.len(), n
                ));
            }
            let sum: f64 = row.iter().sum();
            if (sum - 1.0).abs() > 1e-9 {
                return Err(format!(
                    "transitions row {} sums to {}, expected 1.0",
                    i, sum
                ));
            }
            if row.iter().any(|&p| p < 0.0) {
                return Err(format!(
                    "transitions row {} has negative probability", i
                ));
            }
        }
        for s in &self.states {
            if s.name.is_empty() {
                return Err("state name must be non-empty".into());
            }
            if s.rtp_component < 0.0 {
                return Err("rtp_component must be ≥ 0".into());
            }
        }
        Ok(())
    }
}

/// Solve π × P = π with sum(π) = 1 via Gaussian elimination on (P^T - I).
pub fn stationary_distribution(params: &StateMachineParams) -> Result<Vec<f64>, String> {
    let n = params.states.len();
    let mut a: Vec<Vec<f64>> = vec![vec![0.0; n]; n];
    let mut b: Vec<f64> = vec![0.0; n];

    // Rows 0..n-1: (P^T - I) π = 0
    for i in 0..n - 1 {
        for j in 0..n {
            a[i][j] = params.transitions[j][i];
        }
        a[i][i] -= 1.0;
    }
    // Last row: sum(π) = 1
    for j in 0..n {
        a[n - 1][j] = 1.0;
    }
    b[n - 1] = 1.0;

    // Gaussian elimination with partial pivoting
    for i in 0..n {
        let mut pivot_row = i;
        let mut max_val = a[i][i].abs();
        for k in (i + 1)..n {
            if a[k][i].abs() > max_val {
                max_val = a[k][i].abs();
                pivot_row = k;
            }
        }
        if max_val < 1e-15 {
            return Err("Transition matrix is singular".into());
        }
        a.swap(i, pivot_row);
        b.swap(i, pivot_row);
        for k in (i + 1)..n {
            let factor = a[k][i] / a[i][i];
            for j in i..n {
                a[k][j] -= factor * a[i][j];
            }
            b[k] -= factor * b[i];
        }
    }

    // Back-substitution
    let mut pi: Vec<f64> = vec![0.0; n];
    for i in (0..n).rev() {
        let mut s = b[i];
        for j in (i + 1)..n {
            s -= a[i][j] * pi[j];
        }
        pi[i] = s / a[i][i];
    }

    // Numerical cleanup
    for p in pi.iter_mut() {
        *p = p.max(0.0);
    }
    let total: f64 = pi.iter().sum();
    if total <= 0.0 {
        return Err("stationary distribution sums to ≤ 0".into());
    }
    for p in pi.iter_mut() {
        *p /= total;
    }
    Ok(pi)
}

#[derive(Debug, Serialize)]
pub struct PerState {
    pub name: String,
    pub stationary_probability: f64,
    pub rtp_component_in_state: f64,
    pub weighted_rtp_contribution: f64,
}

#[derive(Debug, Serialize)]
pub struct StateMachineResult {
    pub rtp_contribution: f64,
    pub states_count: u32,
    pub stationary_distribution: Vec<f64>,
    pub states: Vec<PerState>,
}

pub fn state_machine_rtp(p: &StateMachineParams) -> StateMachineResult {
    let pi = stationary_distribution(p).expect("validated already");
    let mut per_state: Vec<PerState> = Vec::with_capacity(p.states.len());
    let mut total = 0.0_f64;
    for (s, &prob) in p.states.iter().zip(pi.iter()) {
        let contrib = prob * s.rtp_component;
        per_state.push(PerState {
            name: s.name.clone(),
            stationary_probability: prob,
            rtp_component_in_state: s.rtp_component,
            weighted_rtp_contribution: contrib,
        });
        total += contrib;
    }
    StateMachineResult {
        rtp_contribution: total,
        states_count: p.states.len() as u32,
        stationary_distribution: pi,
        states: per_state,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn two_state_symmetric() {
        let p = StateMachineParams {
            states: vec![
                GameState { name: "a".into(), rtp_component: 0.9 },
                GameState { name: "b".into(), rtp_component: 1.0 },
            ],
            transitions: vec![vec![0.5, 0.5], vec![0.5, 0.5]],
        };
        let pi = stationary_distribution(&p).unwrap();
        assert!((pi[0] - 0.5).abs() < 1e-10);
        assert!((pi[1] - 0.5).abs() < 1e-10);
    }

    #[test]
    fn supermeter_asymmetric() {
        let p = StateMachineParams {
            states: vec![
                GameState { name: "base".into(), rtp_component: 0.96 },
                GameState { name: "super".into(), rtp_component: 2.50 },
            ],
            transitions: vec![vec![0.99, 0.01], vec![0.50, 0.50]],
        };
        let r = state_machine_rtp(&p);
        // π_base ≈ 0.9804, π_super ≈ 0.0196
        // RTP = 0.9804 × 0.96 + 0.0196 × 2.5 ≈ 0.9902
        assert!((r.rtp_contribution - 0.9902).abs() < 1e-4);
    }
}
